File: codesignal-prep/go/concurrency/bufferedchannels/bufferedchannels.go
// Package bufferedchannels drills the one idea buffered channels give
// you for free: a fixed-capacity, concurrency-safe FIFO queue with
// blocking-when-full and blocking-when-empty built in. The buffer
// capacity *is* the bound.
package bufferedchannels

// BoundedQueue is a fixed-capacity FIFO queue built directly on a
// buffered channel.
type BoundedQueue struct {
	items chan int
}

func NewBoundedQueue(capacity int) *BoundedQueue {
	panic("not implemented")
}

// Push blocks until there is room in the queue.
func (q *BoundedQueue) Push(item int) {
	panic("not implemented")
}

// TryPush returns false immediately if the queue is at capacity, instead
// of blocking. (Hint: select with a default case.)
func (q *BoundedQueue) TryPush(item int) bool {
	panic("not implemented")
}

// Pop blocks until an item is available.
func (q *BoundedQueue) Pop() int {
	panic("not implemented")
}

File: codesignal-prep/python/concurrency/bufferedchannels/bufferedchannels.py
"""A Go buffered channel's behavior -- bounded capacity, Push blocks when
full, Pop blocks when empty -- is exactly what queue.Queue(maxsize=N)
already gives you for free. This exercise is about using it correctly,
including the non-blocking variant: Go's `select` with a `default` case
becomes Python's put_nowait()/get_nowait() plus catching queue.Full /
queue.Empty.
"""
from __future__ import annotations

import queue


class BoundedQueue:
    def __init__(self, capacity: int) -> None:
        self._items: "queue.Queue[int]" = queue.Queue(maxsize=capacity)

    def push(self, item: int) -> None:
        """Blocks until there is room in the queue."""
        raise NotImplementedError

    def try_push(self, item: int) -> bool:
        """Returns False immediately if the queue is at capacity, instead
        of blocking.
        """
        raise NotImplementedError

    def pop(self) -> int:
        """Blocks until an item is available."""
        raise NotImplementedError
